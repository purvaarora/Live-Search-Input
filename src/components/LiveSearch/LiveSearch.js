import React from 'react';
import { isEmpty, get, find } from 'lodash';

import './LiveSearch.scss';


class LiveSearch extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            managersList: [],
            showListMenu: false,
            inputString: '',
            filteredList: []
        }
        this.onFocusHandler = this.onFocusHandler.bind(this);
        this.onBlurHandler = this.onBlurHandler.bind(this);
        this.onClickHandler = this.onClickHandler.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleListItemClick = this.handleListItemClick.bind(this);
    }

    async componentDidMount() {
        // Fetch data from given API and update in state
        const res = await fetch("https://gist.githubusercontent.com/daviferreira/41238222ac31fe36348544ee1d4a9a5e/raw/5dc996407f6c9a6630bfcec56eee22d4bc54b518/employees.json");
        const result = await res.json();
        if (!isEmpty(result)) {
            let tempList = []
            result.data.map((item) => {
                tempList.push({
                    name: get(item, 'attributes.name', ''),
                    email: get(find(result.included, (included) => included.id === get(item, 'relationships.account.data.id', '')), 'attributes.email', ''),
                    id: item.id
                })
            })
            this.setState({
                managersList: tempList,
                filteredList: tempList
            })
        }
    }

    componentDidUpdate() {
        var listItems = document.querySelectorAll(".liveSearch-list li");

        if (!isEmpty(listItems)) {
            var currentLI = 0;

            // remove highlighted class from all list items
            for (var i = 0; i < listItems.length; i++) {
                listItems[i].classList.remove("highlight");
            }

            // Initialize first li as the selected (focused) one:
            listItems[currentLI].classList.add("highlight");

            // Set up a key event handler for the document
            document.addEventListener("keydown", function (event) {
                // Check for up/down key presses

                switch (event.keyCode) {
                    case 38: // Up arrow    
                        // Remove the highlighting from the previous element
                        listItems[currentLI].classList.remove("highlight");

                        currentLI = currentLI > 0 ? --currentLI : 0;     // Decrease the counter      
                        listItems[currentLI].classList.add("highlight"); // Highlight the new element
                        break;
                    case 40: // Down arrow
                        // Remove the highlighting from the previous element
                        listItems[currentLI].classList.remove("highlight");

                        currentLI = currentLI < listItems.length - 1 ? ++currentLI : listItems.length - 1; // Increase counter 
                        listItems[currentLI].classList.add("highlight");       // Highlight the new element
                        break;
                    case 13:
                        this.handleListItemClick(listItems[currentLI].id.split('-')[1])

                }

                let listDiv = document.getElementById('managersList');

                if (listDiv) {
                    let elHeight = listItems[currentLI].offsetHeight;
                    let scrollTop = listDiv.scrollTop;

                    let viewport = scrollTop + listDiv.offsetHeight;
                    var elOffset = elHeight * currentLI;

                    if (elOffset < scrollTop || (elOffset + elHeight) > viewport)
                        listDiv.scrollTop = elOffset;
                }
                listDiv.addEventListener('mouseenter', function (event) {
                    event.target.classList.add('highlight');
                }).addEventListener('mouseleave', function (event) {
                    event.target.removeClass('highlight');
                });

            })
        }
    }

    onFocusHandler() {
        // Shows list on input focus
        this.setState({
            showListMenu: true
        })
    }
    onBlurHandler() {
        // Hides list on input blur 
        setTimeout(() => {
            this.setState({
                showListMenu: false
            })
        }, 500)
    }
    onClickHandler() {
        // handles click on input field
        this.setState({
            showListMenu: !this.state.showListMenu
        })
    }

    handleInputChange(e) {
        // handles input field change and sets entered the value to state
        if (!isEmpty(e.target)) {
            this.setState({
                inputString: e.target.value,
                showListMenu: true,
                filteredList: this.state.managersList.filter((item) => item.name.replace(/\s/g, '').toLowerCase().includes(e.target.value.replace(/\s/g, '').toLowerCase()))
            })

        }
    }

    handleListItemClick(id) {
        // handles enter press on list item and sets input string to state
        this.setState({
            inputString: get(find(this.state.managersList, ['id', id]), 'name', ''),
            showListMenu: false
        })
    }

    render() {
        const { inputString, filteredList, showListMenu } = this.state;
        return (
            <div className='liveSearch'>
                <input
                    type='text'
                    id='liveSearchInput'
                    placeholder='Choose Manager'
                    className='liveSearch-input'
                    value={inputString}
                    onBlur={this.onBlurHandler}
                    onFocus={this.onFocusHandler}
                    onMouseDown={this.onClickHandler}
                    onChange={this.handleInputChange}
                />
                {
                    showListMenu && <ul className='liveSearch-list' id="managersList">
                        {
                            filteredList.map((item) => <li
                                className='liveSearch-list-item'
                                onClick={() => this.handleListItemClick(item.id)}
                                id={`listItem-${item.id}`}
                                key={item.id}
                                onMouseEnter={(e) => { e.target.classList.add("highlight") }}
                                onMouseLeave={(e) => { e.target.classList.remove("highlight") }}
                            >
                                <p className='liveSearch-list-item-acronym'>{item.name.split(/\s/).reduce((response, word) => response += word.slice(0, 1), '')}</p>
                                <div className='liveSearch-list-item-content'>
                                    <p className='liveSearch-list-item-content-name'>{item.name}</p>
                                    <p className='liveSearch-list-item-content-email'>{item.email}</p>
                                </div>
                            </li>)
                        }
                    </ul>
                }
            </div>
        )
    }
}

export default LiveSearch;

