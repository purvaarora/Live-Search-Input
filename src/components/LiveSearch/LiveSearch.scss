html {
    box-sizing: border-box;
}
*, *:before, *:after {
    box-sizing: inherit;
}


.liveSearch {
    text-align: left;
    width: 400px;
    margin: 20px 50px;
    &-input {
        padding: 20px;
        width: 100%;
    }
    &-list {
        height: 140px;
        overflow-y: scroll;
        padding: 0;
        margin: 0;
        &-item {
            border: 1px solid;
            padding: 10px 20px;
            display: flex;
            height: 70px;
            &-acronym {
                padding: 10px;
                background-color: #ffa31a;
                margin: 0;
                color: #fff;
                font-weight: bold;
                line-height: 26px;
            }
            &-content {
                text-align: left;
                margin-left: 20px;
                p {
                    margin: 5px 0;
                }
                &-name {
                    font-weight: bold;
                }
                &-email {
                    font-size: 12px;
                    color: #8c8c8c;
                }
            }
        }
    }
}
.highlight {
    background-color: #e6ffff;
    .liveSearch-list-item-acronym {
        background-color: #66b3ff;
    }
    .liveSearch-list-item-content-name {
        color: #5cd65c;
    }
}