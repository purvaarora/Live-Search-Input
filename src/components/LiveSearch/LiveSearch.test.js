import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
import LiveSearch from './LiveSearch'

const mockResponse = {
    "data": [
        {
            "type": "employees",
            "id": "323",
            "links": {
                "self": "http://localhost:3000/v1/employees/323"
            },
            "attributes": {
                "identifier": null,
                "firstName": "Harriet",
                "lastName": "McKinney",
                "name": "Harriet McKinney",
                "features": [
                    "engagement"
                ],
                "avatar": null,
                "employmentStart": "2016-01-31T00:00:00.000Z",
                "external": false,
                "Last Year Bonus": 3767,
                "Business Unit": "Sales",
                "Commute Time": 34,
                "Age": "1984-02-08",
                "Department": "Customer Care",
                "Gender": "Female",
                "Job Level": "Manager",
                "Local Office": "Kuala Lumpur",
                "% of target": 88,
                "Region": "APAC",
                "Salary": 76000,
                "Tenure": "2014-05-31"
            },
            "relationships": {
                "company": {
                    "data": {
                        "type": "companies",
                        "id": "5"
                    }
                },
                "account": {
                    "data": {
                        "type": "accounts",
                        "id": "324"
                    }
                },
                "phones": {
                    "data": []
                },
                "Manager": {
                    "data": {
                        "type": "employees",
                        "id": "201"
                    }
                }
            }
        },
        {
            "type": "employees",
            "id": "142",
            "links": {
                "self": "http://localhost:3000/v1/employees/142"
            },
            "attributes": {
                "identifier": null,
                "firstName": "Mathilda",
                "lastName": "Summers",
                "name": "Mathilda Summers",
                "features": [
                    "engagement"
                ],
                "avatar": null,
                "employmentStart": "2016-01-31T00:00:00.000Z",
                "external": false,
                "Last Year Bonus": 95050,
                "Business Unit": "Marketing",
                "Commute Time": 131,
                "Age": "1976-12-19",
                "Department": "Media",
                "Gender": "Female",
                "Job Level": "Executive",
                "Local Office": "London",
                "% of target": 166,
                "Region": "EMEA",
                "Salary": 248000,
                "Tenure": "2000-05-15"
            },
            "relationships": {
                "company": {
                    "data": {
                        "type": "companies",
                        "id": "5"
                    }
                },
                "account": {
                    "data": {
                        "type": "accounts",
                        "id": "143"
                    }
                },
                "phones": {
                    "data": []
                },
                "Manager": {
                    "data": {
                        "type": "employees",
                        "id": "139"
                    }
                }
            }
        },
    ],
    "included": [
        {
            "type": "accounts",
            "id": "143",
            "links": {
                "self": "http://localhost:3000/v1/accounts/143"
            },
            "attributes": {
                "email": "mathilda.summers@kinetar.com",
                "locale": null,
                "timezone": null,
                "bouncedAt": null,
                "bounceReason": null,
                "localeEffective": null,
                "timezoneEffective": null
            }
        },
        {
            "type": "accounts",
            "id": "324",
            "links": {
                "self": "http://localhost:3000/v1/accounts/324"
            },
            "attributes": {
                "email": "harriet.mckinney@kinetar.com",
                "locale": null,
                "timezone": null,
                "bouncedAt": null,
                "bounceReason": null,
                "localeEffective": null,
                "timezoneEffective": null
            }
        },
    ],
};

beforeAll(() => {
    global.fetch = jest.fn();
    //window.fetch = jest.fn(); if running browser environment
});


describe('LiveSearch component tests', () => {
    it('matches snapshot', () => {
        const component = renderer.create(
            <LiveSearch />,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('data is fetched on componentDidMount and state is updated', () => {
        it('fetches data from api', async () => {
            const wrapper = shallow(<LiveSearch />)
            const spyDidMount = jest.spyOn(LiveSearch.prototype, "componentDidMount");

            fetch.mockImplementation(() => {
                return Promise.resolve({
                    status: 200,
                    json: () => {
                        return mockResponse;
                    }
                });
            });

            const didMount = wrapper.instance().componentDidMount();
            expect(spyDidMount).toHaveBeenCalled();

            didMount.then(() => {
                // updating the wrapper
                wrapper.update();
                expect(wrapper.instance().state.toEqual({
                    managersList: [
                        {
                            name: 'Harriet McKinney',
                            email: 'harriet.mckinney@kinetar.com',
                            id: '323'
                        },
                        {
                            name: 'Mathilda Summers',
                            email: 'mathilda.summers@kinetar.com',
                            id: '142'
                        }
                    ],
                    showListMenu: false,
                    inputString: '',
                    filteredList: [
                        {
                            name: 'Harriet McKinney',
                            email: 'harriet.mckinney@kinetar.com',
                            id: '323'
                        },
                        {
                            name: 'Mathilda Summers',
                            email: 'mathilda.summers@kinetar.com',
                            id: '142'
                        }
                    ]
                }));

                spyDidMount.mockRestore();
                fetch.mockClear();
                done();
            });


        })

    })

    describe('test state update functions', () => {
        let wrapper;
        beforeAll(() => {
            wrapper = shallow(<LiveSearch />);
            wrapper.setState({
                managersList: [
                    {
                        name: 'Harriet McKinney',
                        email: 'harriet.mckinney@kinetar.com',
                        id: '323'
                    },
                    {
                        name: 'Mathilda Summers',
                        email: 'mathilda.summers@kinetar.com',
                        id: '142'
                    }
                ],
                showListMenu: false,
                inputString: '',
                filteredList: [
                    {
                        name: 'Harriet McKinney',
                        email: 'harriet.mckinney@kinetar.com',
                        id: '323'
                    },
                    {
                        name: 'Mathilda Summers',
                        email: 'mathilda.summers@kinetar.com',
                        id: '142'
                    }
                ]
            });
            wrapper.update();
        })

        it('onFocusHandler shows the dropdownList and onBlurHandler hides the dropdownList', () => {
            expect(wrapper.state("showListMenu")).toEqual(false);

            const onFocusHandlerSpy = jest.spyOn(wrapper.instance(), "onFocusHandler");
            const onBlurHandlerSpy = jest.spyOn(wrapper.instance(), "onBlurHandler");
            wrapper.instance().onFocusHandler();
            expect(onFocusHandlerSpy).toHaveBeenCalledTimes(1);
            expect(wrapper.state("showListMenu")).toEqual(true);

            jest.useFakeTimers();
            wrapper.instance().onBlurHandler();
            jest.advanceTimersByTime(1000);
            expect(onBlurHandlerSpy).toHaveBeenCalledTimes(1);
            expect(wrapper.state("showListMenu")).toEqual(false);
            jest.useRealTimers();

            onFocusHandlerSpy.mockReset();
            onFocusHandlerSpy.mockRestore();
            onBlurHandlerSpy.mockReset();
            onBlurHandlerSpy.mockRestore();

        })

        it('onClickHandler toggles the dropdownList', () => {
            expect(wrapper.state("showListMenu")).toEqual(false);

            const onClickHandlerSpy = jest.spyOn(wrapper.instance(), "onClickHandler");

            wrapper.instance().onClickHandler();
            expect(onClickHandlerSpy).toHaveBeenCalledTimes(1);
            expect(wrapper.state("showListMenu")).toEqual(true);

            wrapper.instance().onClickHandler();
            expect(onClickHandlerSpy).toHaveBeenCalledTimes(2);
            expect(wrapper.state("showListMenu")).toEqual(false);

            onClickHandlerSpy.mockReset();
            onClickHandlerSpy.mockRestore();

        });

        it('handleInputChange updates the input', () => {
            const handleInputChangeSpy = jest.spyOn(wrapper.instance(), "handleInputChange");
            wrapper.instance().handleInputChange({ target: { value: 'har' } });
            expect(handleInputChangeSpy).toHaveBeenCalledTimes(1);
            expect(wrapper.state("showListMenu")).toEqual(true);
            expect(wrapper.state("inputString")).toEqual('har');
            expect(wrapper.state("filteredList")).toEqual([
                {
                    name: 'Harriet McKinney',
                    email: 'harriet.mckinney@kinetar.com',
                    id: '323'
                }
            ]);

            handleInputChangeSpy.mockReset();
            handleInputChangeSpy.mockRestore();

        })

        it('handleListItemClick selects the list element', () => {
            const handleListItemClickSpy = jest.spyOn(wrapper.instance(), "handleListItemClick");

            wrapper.instance().handleListItemClick('142');
            expect(handleListItemClickSpy).toHaveBeenCalledTimes(1);
            expect(wrapper.state("showListMenu")).toEqual(false);
            expect(wrapper.state("inputString")).toEqual('Mathilda Summers');

            handleListItemClickSpy.mockReset();
            handleListItemClickSpy.mockRestore();

        })
        it('trigger click, mouseEnter and mouseLeave for a list item', () => {
            wrapper.find('#liveSearchInput').simulate('change', { target: { value: 'a' } });
            expect(wrapper.find('.liveSearch-list').exists()).toBeTruthy();

            expect(wrapper.find('.liveSearch-list-item').length).toBe(2);
            wrapper.find('.liveSearch-list-item').first().simulate('click', '323');

            wrapper.find('#liveSearchInput').simulate('change', { target: { value: '' } });
            wrapper.find('.liveSearch-list-item').first().simulate('mouseEnter', { target: { classList: { contains: jest.fn(), remove: jest.fn(), add: jest.fn() } } });
            wrapper.find('.liveSearch-list-item').first().simulate('mouseLeave', { target: { classList: { contains: jest.fn(), remove: jest.fn(), add: jest.fn() } } });

        });
    })
})




