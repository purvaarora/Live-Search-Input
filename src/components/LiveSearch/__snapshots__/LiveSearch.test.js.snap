// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveSearch component tests matches snapshot 1`] = `
<div
  className="liveSearch"
>
  <input
    className="liveSearch-input"
    id="liveSearchInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    placeholder="Choose Manager"
    type="text"
    value=""
  />
</div>
`;
